Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +21,45 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return AmpHrUsed;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if(current_t > 0)
+        return true;
+    else
+        return false;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    QTime time(0,0,0,0);
+    int time_left;
+    if(isCharging()){
+        time_left = (totalAmpHoursUsed()/ current_t) * HOUR_TO_MILLISECONDS;
+    }
+    else
+    {
+        time_left = ((BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed())/current_t) * HOUR_TO_MILLISECONDS;
+    }
+
+    return time.addMSecs(time_left);
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
     Q_UNUSED(batteryData);
-    // This is where you can update your variables
-    // Hint: There are many different ways that the totalAmpHoursUsed can be updated
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+
+    current_t = batteryData.current;
+    voltage_t = batteryData.voltage;
+    time_t = batteryData.time;
+
+    prevCurr = current_t;
```

Comment:
Remaining `current_t` to something like `presentCurrent` would go better with `previousCurrent`